internal/netTestutils.go

```diff
@@ -0,0 +1,144 @@
+// external network.
+type LocalNetworkServices interface {
+	// DNSServerIP returns the IP address of the DNS server
+	// (useful for seting resolv.conf)
+	DNSServerIP() string
+
+	// URL returns the full URL to use when querying the provided
+	// subpath. i.e. if subpath is "foo/bar", this will return
+	// something like "http://phony.domain.test/foo/bar"
+	URL(subpath string) string
+
+	// Serve will start the DNS and HTTP servers, blocking until
+	// either of them shuts down.
+	Serve(ctx context.Context) error
+}
+
+// NewLocalNetworkServices returns an implementation of
+// LocalNetworkServices where the DNS server will bind to the
+// ip of a created device and serve a single mapping from
+// "domain" to that ip. The HTTP server will serve a pages at
+// paths defined in the keys of "webpages", with the content set
+// in the values of that map.
+func NewLocalNetworkServices(webpages map[string]string) (LocalNetworkServices, error) {
+	testDevName := "testdev0"
+	ipAddr := "10.0.0.1"
+	ipCidr := fmt.Sprintf("%s/32", ipAddr)
+
+	output, err := exec.Command("ip", "tuntap", "add", testDevName, "mode", "tun").CombinedOutput()
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to add tun dev, \"ip\" output: %s", string(output))
+	}
+
+	output, err = exec.Command("ip", "addr", "add", ipCidr, "dev", testDevName).CombinedOutput()
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to assign ip to tun dev, \"ip\" output: %s", string(output))
+	}
+
+	output, err = exec.Command("ip", "link", "set", "dev", testDevName, "up").CombinedOutput()
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to set tun dev up, \"ip\" output: %s", string(output))
+	}
+
+	return &localNetworkServices{
+		domain:   "phony.domain.test",
+		webpages: webpages,
+		ipAddr:   ipAddr,
+	}, nil
+}
+
+type localNetworkServices struct {
+	domain   string
+	webpages map[string]string
+	ipAddr   string
+}
+
+func (l localNetworkServices) DNSServerIP() string {
+	return l.ipAddr
+}
+
+func (l localNetworkServices) URL(subpath string) string {
+	return fmt.Sprintf("http://%s/%s", l.domain, subpath)
+}
+
+func (l localNetworkServices) Serve(ctx context.Context) error {
+	errGroup, _ := errgroup.WithContext(ctx)
+
+	errGroup.Go(func() error {
+		dnsSrv := &dns.Server{
+			Addr: l.ipAddr + ":53",
+			Net:  "udp",
+			Handler: &dnsHandler{
+				records: map[string]string{
+					l.domain + ".": l.ipAddr,
+				},
+			},
+		}
+		return dnsSrv.ListenAndServe()
+	})
+
+	errGroup.Go(func() error {
+		for path, contents := range l.webpages {
+			webpage := contents
+			http.HandleFunc("/"+path, func(w http.ResponseWriter, r *http.Request) {
+				fmt.Fprintf(w, webpage)
+			})
+		}
+
+		return http.ListenAndServe(l.ipAddr+":80", nil)
+	})
+
+	return errGroup.Wait()
+}
+
+type dnsHandler struct {
+	records map[string]string
+}
+
+func (h dnsHandler) ServeDNS(w dns.ResponseWriter, r *dns.Msg) {
+	msg := dns.Msg{}
+	msg.SetReply(r)
+	if r.Question[0].Qtype == dns.TypeA {
```

Comment:
Does this return `NXDOMAIN` if it's not a question for an `A` record?

Reply:
`dig` should give you more clear answers for DNS-specific questions than `wget`

Reply:
I'll give that a try

Reply:
Tried it, it was not failing with `NXDOMAIN` specifically, so I updated the code to now return `NXDOMAIN` if the name is not found. I don't think it makes a difference at the moment but good catch either way, it may matter in the future.